### ISSUE TITLE
Single-ELF (fused OperatorSequence) dispatch intermittently returns stale data at high-magnitude channels; ERT reports COMPLETED

### ISSUE THREAD

**Issue by bong-water-water-bong** [Xilinx/mlir-aie#3656]
## Summary

A single-ELF fused FFN (6 ops in one `OperatorSequence` dispatch: gate GEMM -> up GEMM -> SiLU -> eltwise mul -> down GEMM -> residual add) intermittently returns data where the FFN contribution is MISSING at the model's high-magnitude channels, while the ERT reports COMPLETED. Same inputs, same ELF, repeated dispatches: ~1-in-3 runs drop the FFN output at those channels (reads back ~0 there); the rest match a bf16-faithful CPU reference to 1 ULP.

## Environment

- Ryzen AI MAX+ 395 (Strix Halo), XDNA2 NPU (npu2), amdxdna driver, XRT 2.21.75
- amd/iron (devel @ 6410550), mlir-aie aiecc full-ELF path (`OperatorSequence`, dispatch="auto" -> fused)
- Llama-3.2-1B prefill buffers (M=2048, K=2048/8192, N=8192/2048, bf16, 8 columns)

## Reproduction (fused_warmup.py, in the issue's linked gist/repo)

Fused runlist (ONE hw_context): `(gate GEMM, x_norm, W_ffn_gate, ffn_gate) -> (up GEMM, ..., ffn_up) -> (SiLU, ffn_gate, ffn_gate) -> (eltwise mul, ffn_gate, ffn_up, ffn_hidden) -> (down GEMM, ffn_hidden, W_ffn_down, ffn_output) -> (residual add, x, ffn_output, x)` with shared weight buffers, invoked once per layer with layer L's weights written between calls. Inputs: layer-0 post-attention x and post-RMSNorm x_norm (CPU-verified). Reference: CPU post-FFN residual x.

After a warm-up dispatch, three identical dispatches:

| run | max diff vs CPU ref | bad elements (>0.5) |
|---|---|---|
| 0 | 0.125 | 0 |
| 1 | 0.125 | 0 |
| 2 | 8.21 | 3 (dims 636, 735, 762) |

The failing runs read `x == x_attn` at the affected channels, i.e. `ffn_output` reads back as ~0 there while the correct value is ~2-24 in magnitude. The affected channels are the same set every time a miss occurs — the model's high-magnitude FFN-output channels (e.g. layer 0 dims 636, 735, 762, 841, 894, 1002, 1314, 1334, 1476, 1503, 1619, 1703, 2037; |ffn_out| ~ 2-24). Everything else matches the reference within 1 ULP.

## What was ruled out

- **Kernel arithmetic**: each op isolated with bf16-faithful references (bf16 inputs, fp32 accumulate, bf16 out) is exact: gate GEMM max 0.094; down GEMM max 1 ULP (summation-order noise only); eltwise mul exact.
- **bf16 precision**: CPU bf16 vs CPU fp32 logits correlate 0.9998 — the model is not bf16-fragile.
- **Host-side buffer handling**: the same inputs produce correct results on most dispatches; the miss is per-dispatch, not per-input.
- **Rebind/BO-churn** (#3653 family): the venv runtime rebind now syncs device->host before BO replacement and binds each tensor to one group; the miss persists (it is inside the single-ELF dispatch itself).

## Notes

- The miss appears to target high-magnitude values: when the FFN output at a channel is large (|~2-24|), a miss reads ~0 there; when the same dispatch's other channels are small, they are correct.
- A synthetic repro at small shapes (M=256, K=256/1024, N=512/1024, bf16, 8 cols) does NOT show the intermittent big-channel miss; it instead shows a consistent ~1% of elements zeroed at scattered positions. The two may share a root cause in the DMA data-mover layer.
- Related: #3654, #3655 (multi-context execution silently returns garbage/COMPLETED), #3594 (hw context residency).

## Ask

Investigate data-mover / tile-state handling for high-magnitude payloads in single-ELF (fused) dispatches: why does a repeated dispatch with identical inputs occasionally not write the full output (ERT COMPLETED, ~0 at specific channels), and why do the affected channels correlate with output magnitude?

**Comment by bong-water-water-bong** [comment]
Full repro script (fused_warmup.py) — requires amd/iron + Llama-3.2-1B weights + Strix Halo NPU2. It builds the fused FFN ELF, warms up, then runs 3+ identical dispatches printing the max deviation vs a CPU reference:

```python
#!/usr/bin/env python3
"""Fused FFN: 6-op single-ELF dispatch, repeated on identical inputs.
Run from iron/applications/llama_3.2_1b with PYTHONPATH=/usr/lib/python3/dist-packages
(iron-venv python). Needs llama3.2-1b model.safetensors + tokenizer.model.
"""
import sys, numpy as np, torch
sys.path.insert(0, "/home/bcloud/amd-oss/iron")
from llama_inference_harness import LlamaConfig
from iron.common.context import AIEContext
from iron.common.sequence import OperatorSequence
from iron.operators import GEMM, SiLU, ElementwiseMul, ElementwiseAdd

PT, EMB, HID = 2048, 2048, 8192
elf_ctx = AIEContext(build_dir="build_elf_ffn")
g = GEMM(M=PT,K=EMB,N=HID,tile_m=64,tile_k=64,tile_n=64,num_aie_columns=8,b_col_maj=False,context=elf_ctx)
u = GEMM(M=PT,K=EMB,N=HID,tile_m=64,tile_k=64,tile_n=64,num_aie_columns=8,b_col_maj=False,context=elf_ctx)
d_ = GEMM(M=PT,K=HID,N=EMB,tile_m=64,tile_k=64,tile_n=64,num_aie_columns=8,b_col_maj=False,context=elf_ctx)
s = SiLU(size=PT*HID,tile_size=HID,num_aie_columns=8,context=elf_ctx)
m = ElementwiseMul(size=PT*HID,tile_size=HID,num_aie_columns=8,context=elf_ctx)
r = ElementwiseAdd(size=PT*EMB,tile_size=EMB,num_aie_columns=8,context=elf_ctx)
runlist = [(g,"x_norm","W_ffn_gate","ffn_gate"),(u,"x_norm","W_ffn_up","ffn_up"),
           (s,"ffn_gate","ffn_gate"),(m,"ffn_gate","ffn_up","ffn_hidden"),
           (d_,"ffn_hidden","W_ffn_down","ffn_output"),(r,"x","ffn_output","x")]
seq = OperatorSequence("ffn_fused",runlist,
    input_args=["x_norm","W_ffn_gate","W_ffn_up","W_ffn_down","x"],
    output_args=["x"], context=elf_ctx).compile()
call = seq.get_callable()
config = LlamaConfig(sys.argv[1], sys.argv[2])
for name, w in [("W_ffn_gate",'model.layers.0.mlp.gate_proj.weight'),
                ("W_ffn_up",'model.layers.0.mlp.up_proj.weight'),
                ("W_ffn_down",'model.layers.0.mlp.down_proj.weight')]:
    call.get_buffer(name).torch_view()[:] = config.weights[w].T.contiguous().flatten()
# inputs: CPU post-attention x + post-RMSNorm x_norm for layer 0; ref: CPU post-FFN x
# (computed by llama_cpu.py; per-layer dumps). Then, after a warm-up call:
#   call(); out = call.get_buffer("x").to_torch().float().numpy()[:7*EMB].reshape(7,EMB)
# Expect: runs 0-1 max 0.125 (correct), run 2 max ~8.2 with ffn missing at dims
# 636/735/762/... (the high-magnitude channels), ERT COMPLETED throughout.
```

Observed on Strix Halo (Ryzen AI MAX+ 395), XRT 2.21.75, amd/iron devel @ 6410550.

**Comment by bong-water-water-bong** [comment]
### Update: magnitude-dependent corruption, localized to the fused ELF's internal data path

Further investigation (synthetic fused FFN, M=256 K=256/1024 N=512/1024, 8 cols, bf16; same runlist structure) shows the corruption is **monotonic in output magnitude** and lives in the fused ELF's internal (on-chip ObjectFIFO) path:

| max |output| | elements wrong (>1.0) |
|---|---|---|
| 0.49 | 0.00% |
| 1.59 | 0.00% |
| 10.3 | 1.92% |
| 41.0 | 23.01% |
| 181 | 54.71% |
| 1112 | 81.04% |
| 3920 | 90.49% |

(8 dispatches each; threshold 1.0 absolute on outputs scaled by the input scale factor.)

Localization within the fused chain:
- `ffn_up` (up-projection GEMM output, read back) is exact at |107| — the GEMM compute and its L3 write are fine.
- `ffn_hidden` (eltwise-mul output, read back) is wrong/zeroed at 12-47% of large positions (run-dependent) — the loss appears at/after the mul's internal input path.
- Every op **isolated** (standalone xclbin, L3 in/out) is exact at the same magnitudes, including the mul at |ref|~1000+ (bf16-faithful reference).
- `emulate_bf16_mmul_with_bfp16=False` changes nothing.

So: the fused (single-ELF) dispatch's on-chip data path between op tiles corrupts bf16 payloads with a probability that grows with the payload magnitude, while the same data via the external L3 path is exact. ERT reports COMPLETED throughout. Repro scripts: `mag_curve.py`, `localize_loss.py`, `hidden_selfcheck.py` (small, no model weights; the llama-based `fused_warmup.py` shows the same at the model's high-magnitude channels, ~1-in-3 dispatches).

Practical implication: keep fused-ELF bf16 payloads bounded (|x| < ~2) — e.g., quantize/clip the model's outlier channels — or the drop rate becomes unacceptable at the values llama actually produces (200-600).

**Comment by bong-water-water-bong** [comment]
### Correction + cleaner evidence: intermittent, magnitude-scaled, per-dispatch

Rechecking with a bf16-faithful stage-by-stage reference, a large part of the
earlier "magnitude curve" was rounding-noise amplification through the
synthetic chain (intermediate values explode with input scaling) — that part
is NOT corruption. The clean discriminator is **per-run variance on identical
dispatches**:

synthetic fused FFN, identical inputs, 20 consecutive dispatches:

- scale 2 (max|out| ~1.4): bad elements per run {600..631}, **20/20 runs
  unique outputs** (~0.23% of elements wrong each run, varying).
- scale 5 (max|out| ~11): bad elements per run {54k..89k}, **20/20 runs
  unique outputs** (~21-34% wrong, varying).

Every dispatch with the same inputs returns a different result — genuine
intermittent corruption, rate increasing with payload magnitude. The llama
FFN (values |200-600|) shows the same intermittency (1-in-3 dispatches lose
the FFN contribution at its high-magnitude channels). The exact channel set
and per-run counts vary; ERT reports COMPLETED every time.

Please treat the earlier fp32-reference magnitude table as superseded by
this per-run-variance evidence.

**Comment by bong-water-water-bong** [comment]
### ROOT CAUSE FOUND — in-place (aliased) buffers in the fused dispatch

The corruption is caused by **a runlist op aliasing a buffer as both input and output (in-place)**. With the SAME ELF, SAME inputs, SAME kernels:

- In-place residual `(residual_add, "x", "ffn_output", "x")`: identical dispatches return different results (per-run variance; 20/20 unique outputs in one run, 3/10 in another), wrong elements at scattered positions.
- Out-of-place residual `(residual_add, "x", "ffn_output", "x_out")`: **15/15 dispatches bit-identical, 0 wrong elements** — deterministic and correct.

The device inputs were verified bit-identical before every dispatch in both cases, so this is not a harness artifact. The "magnitude dependence" reported earlier was the error threshold: the aliasing race produces stale-read errors whose absolute size scales with the payload values, so larger values simply exceed the detection threshold more often.

The affected pattern appears in the full-ELF (fused) dispatch path when the consolidated-buffer offsets for a runlist buffer collide between its input and output roles (the same named buffer used as both an op's input and output). The standalone (separate-xclbin) path with the same in-place add is clean (tested: 0 bad over 15 runs).

Actionable: the fused-ELF dispatch must handle aliased (in-place) buffers correctly — allocate distinct read/write views, or detect and reject the pattern. The decode app's runlist uses the same in-place pattern (`(residual_add_op, "x", "attn_output", "x")`), which is consistent with the decode-phase text variance we also observed.

**Comment by bong-water-water-bong** [comment]
### FIXED (runtime) — in-place buffer writes never reached the device

Root cause of the fused-dispatch intermittency: an in-place runlist buffer (a
name in BOTH input_args and output_args, e.g. `(residual_add, "x",
"ffn_output", "x")`) is allocated in the OUTPUT consolidated arena (the
output allocation overwrites the layout entry), so writing it via
`get_buffer(...).torch_view()` marks the OUTPUT BO dirty — but
`SequenceFullELFCallable._sync_inputs()` only pushed the INPUT BO. The write
never reached the device: x stayed stale/racy, so dispatches intermittently
ran against old x (the "missing FFN contribution" pattern) and the decode
ELF (same in-place pattern) produced run-to-run text variance.

Fix (amd/iron iron/common/sequence.py `_sync_inputs`): push dirty ranges on
all three consolidated buffers (input, output, scratch) before every
dispatch.

Verified: the in-place fused FFN (previously 3/10 unique outputs, bad runs)
is now **20/20 dispatches bit-identical, 0 wrong elements**; the llama
decode is **3/3 runs identical** ("SCENE I. King Leontes and Camillo's
palace."). The prefill logits match the true CPU reference at corr 0.9974
(bf16 tolerance) — the earlier corr-0.32 "divergence" was a polluted CPU
reference (its logits dump was overwritten by the decode pass), not an NPU
error.

Suggest adding a regression test: a fused sequence with an in-place
residual, dispatched repeatedly, asserting bit-identical outputs.

**Comment by bong-water-water-bong** [comment]
Resolved. Root cause: in a single-ELF `OperatorSequence`, a runlist buffer that appears in **both** `input_args` and `output_args` (the residual_add `"x"`) is allocated in the **output** arena (the output allocation overwrites the layout entry). `_sync_inputs` only pushed the input BO, so the in-place write to `"x"` never reached the device — the FFN contribution was missing at high-magnitude channels while ERT reported COMPLETED.

Fix (amd/iron devel, commit `76eee2a`, `iron/common/sequence.py`):
- `SequenceFullELFCallable._sync_inputs` now pushes **all three** consolidated buffers (input, output, scratch) — the host-dirty ranges of in-place buffers in the output arena get to the device.
- Regression test `test_fused_inplace_buffer_bit_stable` (10/10).
- Also fixed the fused read-back shape (flat 1-D `get_buffer("x")` must reshape to `(max_seq_len, emb_dim)`).

Verified on Strix Halo XDNA2: fused prefill FFN **20/20 bit-identical** per layer; llama decode 3/3 identical; prefill logits corr 1.0 vs the separate-op path.
